autode/neb/idpp.py

```diff
@@ -0,0 +1,137 @@
+import numpy as np
+from scipy.spatial import distance_matrix
+
+
+class IDPP:
+    """
+    Image dependent pair potential (IDPP) objective function from
+    https://arxiv.org/pdf/1406.1512.pdf
+
+    .. math::
+
+        S = Σ_i  Σ_{j>i} w(r_{ij}) (r_{ij}^k - r_{ij})^2
+
+    where :math:`r_{ij}` is the distance between atoms i and j and
+    :math:`r_{ij}^k = r_{ij}^0 + k(r_{ij}^N - r_{ij}^0)/N` for :math:`N` images.
+    Note: here the superscripts denote the image number, not a power. The
+    weight function is :math:`w(r_ij) = r_{ij}^{-4}`, as suggested in the paper
+    """
+
+    def __init__(self, images: 'autode.neb.original.Images'):
+        """Initialise a IDPP potential"""
+
+        if len(images) < 2:
+            raise ValueError('Must have at least 2 images for IDPP')
+
+        # Distance matrices containing all r_{ij}^k
+        self._dists = {image.name: None for k, image in enumerate(images)}
+        self._diagonal_distance_matrix_idxs = None
+
+        self._set_distance_matrices(images)
+
+    def __call__(self,
+                 image: 'autode.neb.original.Image'
+                 ) -> float:
+        """Value of the IDPP objective function for a single image
+
+
+        S = 0.5 Σ  Σ  w(r_{ij}) (r_{ij}^k - r_{ij})^2
+                i j≠i
+        """
+        r_k, r = self._req_distance_matrix(image), self._distance_matrix(image)
+
+        return 0.5 * np.sum(self._w(image) * (r_k - r)**2)
+
+    def grad(self,
+             image: 'autode.neb.original.Image'
+             ) -> np.ndarray:
+        r"""
+        Gradient of the potential with respect to displacement of
+        the Cartesian components: :math:`\nabla S = (dS/dx_0, dS/dy_0, dS/dz_0,
+        dS/dx_1, ...)` where the numbers denote different atoms. For example,
+
+        .. math::
+
+            \frac{dS}{dx_0} = -2 \sum_{i \ne j}
+                               \left[2(c-r_{ij})r_{ij}^{-6}
+                                     + w(r_{ij})r_{ij}^{-1})
+                               \right](c - r_{ij})(x_0 - x_j)
+
+        where :math:`c = r_{ij}^k`.
+        """
+
+        n_atoms = image.species.n_atoms
+        x = np.array(image.species.coordinates).flatten()
+        grad = np.zeros(shape=(n_atoms, 3))
+
+        r = self._distance_matrix(image, unity_diagonal=True)
+        w = self._w(image)
+        r_k = self._req_distance_matrix(image)
+
+        a = -2 * (2 * (r_k - r)**2 * r**(-6)
```

Comment:
I'm assuming you've checked all the maths is correct

Reply:
Checked against the numerical in a test, also PFFT - have faith 😉